daemon/fmutex/fmutex.go

```diff
@@ -0,0 +1,81 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2015 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package fmutex implements locking for the snappy daemon
+package fmutex
+
+import (
+	"sync"
+
+	"github.com/ubuntu-core/snappy/dirs"
+	"github.com/ubuntu-core/snappy/logger"
+	"github.com/ubuntu-core/snappy/priv"
+)
+
+// FMutex gives you an exclusive mutex that also holds the filesystem lock.
+type FMutex struct {
+	mutex sync.Locker
+	flock FLocker
```

Comment:
the "I should not test this package?" was because with a concrete value there no tests at all would be possible, even the silly ones it has now, because they'd all fail because tests are not run as root.

I thought the double-locking you were referring to was about holding the lock twice, which the bare filesystem lock (ignoring the root thing) fails.

I've closed this pull request, will push up a branch refactoring priv, and then attempt this one again.